internal/provider/kubernetes/route.go

```diff
@@ -0,0 +1,421 @@
+		found, err := isRoutePresentInNamespace(ctx, r.client, hrkey.Namespace)
+		if err != nil {
+			return requests
+		}
+		if !found {
+			r.resources.Namespaces.Delete(hrkey.Namespace)
+			r.log.Info("deleted namespace from watchable map")
+		}
+
+		// Delete the Service from the watchable maps if no other
+		// routes (TLSRoute or HTTPRoute) reference that Service.
+		routeServices := r.referenceStore.getRouteToServicesMapping(ObjectKindNamespacedName{kindHTTPRoute, hrkey.Namespace, hrkey.Name})
+		for svc := range routeServices {
+			r.referenceStore.removeRouteToServicesMapping(ObjectKindNamespacedName{kindHTTPRoute, hrkey.Namespace, hrkey.Name}, svc)
+			if !r.referenceStore.isServiceReferredByRoutes(svc) {
+				r.resources.Services.Delete(svc)
+				r.log.Info("deleted service from watchable map", "namespace", svc.Namespace, "name", svc.Name)
+			}
+		}
+	}
+
+	if !routeDeleted {
+		v, ok := r.resources.HTTPRoutes.Load(hrkey)
+		// Donot process further if the resource is unchanged.
+		if ok && httproute.Generation == v.Generation {
+			return requests
+		}
+		if !ok {
+			r.resources.HTTPRoutes.Store(hrkey, httproute)
+			r.log.Info("added httproute to watchable map")
+		}
+
+		// Get the route's namespace from the cache.
+		if err := r.checkNamespaceForRoute(ctx, hrkey.Namespace); err != nil {
+			return requests
+		}
+
+		// Get the route's backendRefs from the cache. Note that a Service is the
+		// only supported kind.
+		routeBackendReferences := []gwapiv1b1.BackendRef{}
+		for i := range httproute.Spec.Rules {
+			for j := range httproute.Spec.Rules[i].BackendRefs {
+				ref := httproute.Spec.Rules[i].BackendRefs[j].BackendRef
+				routeBackendReferences = append(routeBackendReferences, ref)
+			}
+		}
+		if err := r.checkAndValidateRouteBackendRefs(ctx, kindHTTPRoute, hrkey, routeBackendReferences); err != nil {
+			return requests
+		}
+
+		httpRouteParentReferences = append(httpRouteParentReferences, httproute.Spec.ParentRefs...)
+	}
+
+	// Find the parent Gateway objects for httproute.
+	gateways, err := validateParentRefs(ctx, r.client, hrkey.Namespace, r.classController, httpRouteParentReferences)
+	if err != nil {
+		r.log.Info("invalid parentRefs for httproute, bypassing reconciliation", "object", obj)
+		return requests
+	}
+
+	// Remove the route from the watchable map since it doesn't reference
+	// a managed Gateway.
+	if len(gateways) == 0 {
+		r.log.Info("httproute doesn't reference any managed gateways")
+		r.resources.HTTPRoutes.Delete(hrkey)
+		return requests
+	}
+
+	for j := range gateways {
+		requests = append(requests, r.processGateway(gateways[j].DeepCopy())...)
+	}
+
+	return requests
+}
+
+// addHTTPRouteIndexers adds indexing on HTTPRoute, for Service objects that are
+// referenced in HTTPRoute objects via `.spec.rules.backendRefs`. This helps in
+// querying for HTTPRoutes that are affected by a particular Service CRUD.
+func addHTTPRouteIndexers(ctx context.Context, mgr manager.Manager) error {
```

Comment:
trying to understand how does the controller watch for specific Objects i.e. at `t0` when an `HTTPRoute`
references Secret `secret-a`, the controller also watches for changes to `secret-a`. However at `t1` when the
same `HTTPRoute` references `secret-b` instead of `secret-a`, does the logic ensure that we stop watching `secret-a`
and watch `secret-b` instead ?

Reply:
I don't think dynamically changing the watch scope would be possible. Although in general, the logic takes care when the service/secret or any other object reference changes, and adds corresponding objects to the resource Map accordingly.